src/ssl.c

```diff
@@ -15243,140 +15277,177 @@ WOLFSSL_SESSION* GetSession(WOLFSSL* ssl, byte* masterSecret,
-            SESSION_ROW_UNLOCK(&SessionCache[lockedRow]);
+        if (ret != BAD_MUTEX_E) {
+            SESSION_ROW_UNLOCK(&SessionCache[row]);
         }
     }
+#endif
 
     if (ret != WOLFSSL_SUCCESS) {
+    #ifdef HAVE_SESSION_TICKET
         /* cleanup */
-        if (ticBuff) {
-            XFREE(ticBuff, ssl->heap, DYNAMIC_TYPE_SESSION_TICK);
+        if (ssl->session.ticket != ssl->session._staticTicket) {
+            XFREE(ssl->session.ticket, ssl->heap, DYNAMIC_TYPE_SESSION_TICK);
         }
-        copyInto->ticket = copyInto->staticTicket;
-        copyInto->isDynamic = 0;
+        ssl->session.ticket = ssl->session._staticTicket;
+        ssl->session.ticketLenAlloc = 0;
+    #endif
+        return ret;
     }
-#endif /* HAVE_SESSION_TICKET */
-
-    return ret;
-}
-
-int SetSession(WOLFSSL* ssl, WOLFSSL_SESSION* session)
-{
-    if (ssl == NULL || SslSessionCacheOff(ssl, session))
-        return WOLFSSL_FAILURE;
+    session = NULL; /* invalidate the provided session, only use ssl->session */
 
 #ifdef OPENSSL_EXTRA
     /* check for application context id */
     if (ssl->sessionCtxSz > 0) {
-        if (XMEMCMP(ssl->sessionCtx, session->sessionCtx, ssl->sessionCtxSz)) {
+        if (XMEMCMP(ssl->sessionCtx, ssl->session.sessionCtx, ssl->sessionCtxSz)) {
             /* context id did not match! */
             WOLFSSL_MSG("Session context did not match");
-            return SSL_FAILURE;
+            return WOLFSSL_FAILURE;
         }
     }
 #endif /* OPENSSL_EXTRA */
 
-    if (LowResTimer() < (session->bornOn + session->timeout)) {
-        int ret = GetDeepCopySession(ssl, session);
-        if (ret == WOLFSSL_SUCCESS) {
-            ssl->options.resuming = 1;
+    if (LowResTimer() < (ssl->session.bornOn + ssl->session.timeout)) {
+        ssl->options.resuming = 1;
 
 #if defined(SESSION_CERTS) || (defined(WOLFSSL_TLS13) && \
                                defined(HAVE_SESSION_TICKET))
-            ssl->version              = session->version;
+        ssl->version              = ssl->session.version;
 #endif
 #if defined(SESSION_CERTS) || !defined(NO_RESUME_SUITE_CHECK) || \
                         (defined(WOLFSSL_TLS13) && defined(HAVE_SESSION_TICKET))
-            ssl->options.cipherSuite0 = session->cipherSuite0;
-            ssl->options.cipherSuite  = session->cipherSuite;
+        ssl->options.cipherSuite0 = ssl->session.cipherSuite0;
+        ssl->options.cipherSuite  = ssl->session.cipherSuite;
 #endif
-        }
-
-        return ret;
+        ret = WOLFSSL_SUCCESS;
     }
     else {
 #if defined(OPENSSL_EXTRA) && defined(WOLFSSL_ERROR_CODE_OPENSSL)
         WOLFSSL_MSG("Session is expired but return success for \
                               OpenSSL compatibility");
-        return WOLFSSL_SUCCESS;
+        ret = WOLFSSL_SUCCESS;
 #endif /* OPENSSL_EXTRA && WOLFSSL_ERROR_CODE_OPENSSL */
-        return WOLFSSL_FAILURE;  /* session timed out */
+        ret = WOLFSSL_FAILURE;  /* session timed out */
```

Comment:
Previously this line was not invoked when the above lines were compiled in.

Reply:
Great eye! Pushed fix.